server/src/main/java/org/elasticsearch/index/snapshots/blobstore/SlicedInputStream.java

```diff
@@ -24,12 +24,12 @@
 public abstract class SlicedInputStream extends InputStream {
     private int nextSlice = 0;
     private InputStream currentStream;
-    private int currentSliceOffset = 0;
+    private long currentSliceOffset = 0;
     private final int numSlices;
     private boolean closed = false;
     private boolean initialized = false;
```

Comment:
Reset when reads beyond EOF have happened cause an assertion error. I did not see this being illegal anywhere and seems to be support for instance by `InputStreamIndexInput` and `ByteArrayInputStream`. I think supporting this requires doing away with the initialized flag and the assertion around it.

It is unlikely that the consumer of this relies upon this behavior, but I'd prefer to fix it.

Test to demonstrate:

```
    public void testResetBeyondEOF() throws IOException {
        SlicedInputStream input = new SlicedInputStream(1) {
            @Override
            protected InputStream openSlice(int slice) throws IOException {
                assertThat(slice, equalTo(0));
                return new ByteArrayInputStream(new byte[] { 1 });
            }
        };
        input.mark(0);

        for (int i = 0; i < 10 ; ++i) {
            System.out.println(i);
            for (int j = 0; j < i; ++j) {
                input.read();
            }
            input.reset();
        }
    }
```

Reply:
Finally extended the assertion in nextStream() as the simplest way.

Reply:
My test snippet was just to demonstrate, happy to have it any other way.